### PR TITLE
solana-ibc: introduce basic interface for handling the guest chain

### PR DESCRIPTION
**Pull request by mina86**
Add contract calls for managing the guest blockchain.  This includes
initialising the chain, generating new blocks, adding signatures and
candidates management.

The implementation is purposefully bare bones with no handling for
rewards or slashing so that the code can be tested and those features
added later on.

### CODE REVIEW

File: solana/solana-ibc/programs/solana-ibc/src/chain.rs [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+            genesis_epoch,
+        )
+        .map_err(|err| Error::Internal(err.into()))?;
+        let manager = Manager::new(config, genesis).map_err(Error::from)?;
+        if self.inner.is_some() {
+            return Err(Error::ChainAlreadyInitialised.into());
+        }
+        let last_check_height = manager.head().1.host_height;
+        self.inner = Some(ChainInner { last_check_height, manager });
+        Ok(())
```
Comment (blasrodri) [inline solana/solana-ibc/programs/solana-ibc/src/chain.rs:47]: maybe we would like to submit an event here? Stating that the chain has started and the hash is 0x*** and any other relevant data?
No info about the val set?

Comment (mina86) [reply solana/solana-ibc/programs/solana-ibc/src/chain.rs:47]: D’oh, yes, this got lost in rebases. I’ll fix it in a moment.

**Comment by dhruvja** [comment]
Should we write some tests too?

**Comment by mina86** [comment]
Yes, though I have been postponing it in interest of rapid development. First priority at the moment is to get something out the doors with something we can demonstrate with local environments on command line.

**Comment by dhruvja** [comment]
> Yes, though I have been postponing it in interest of rapid development. First priority at the moment is to get something out the doors with something we can demonstrate with local environments on command line.

Sure, that makes sense.

**Review by dhruvja — APPROVED** [review]

**Pull request merged**